Various commits to improve consistency in tools

clumens opened the pull request:
On the last commit (b616a49), I am fairly certain but not 100% positive this is okay.

kgaillot commented on tools/crm_node.c:
```diff
@@ -562,10 +561,11 @@ main(int argc, char **argv)
+        g_set_error(&error, PCMK__EXITC_ERROR, exit_code,
+                    "The supplied command is considered dangerous."
+                    "  To prevent accidental destruction of the cluster,"
+                    " the --force flag is required in order to proceed.");
```
Why convert just this and a few other messages to GError? GError obviously makes sense when used with the glib functions that require it, and it makes some sense when we might want to fprintf() or out->err() conditionally, but I don't see any advantage over just fprintf() in other cases. If we did want to use it more, I'd think we want to convert all of the error messages to it.

clumens replied:
I am constantly struggling with how much should use GError, so I don't have a good answer on this.  I can put this one back as fprintf, at least.  The GError usage around pcmk__output_new was mainly because some tools were already doing so.  For g_option_context_parse_strv, command line processing is already using GError so it fits better here.

I've additionally considered converting pcmk__output_new to take a GError to save having to do that block everywhere, but it doesn't seem like it saves much.

kgaillot commented on tools/stonith_admin.c:
```diff
@@ -605,9 +604,18 @@ main(int argc, char **argv)
+        }
+
+        g_clear_error(&error);
+    }
```
This seems used often enough to merit functionizing and using in all tools

clumens replied:
Yeah, great idea.  This is exactly the sort of think I am looking to do in this PR.

kgaillot reviewed (commented):
Nice to see things coming together

clumens commented on lib/common/logging.c:
```diff
@@ -974,3 +974,14 @@ crm_log_output_fn(const char *file, const char *function, int line, int level, c
+        /* These arguments are ignored, so pass placeholders. */
+        crm_bump_log_level(0, NULL);
+    }
+}
```
Fixed, though now that I use it more, it seems weird to pass verbosity to pcmk__cli_init_logging instead of the log level to use.

kgaillot replied:
crm_bump_log_level() is kind of weird, the first time it is called it just enables stderr, and later times call set_crm_log_level() to bump by one. Practically that means it takes one extra -V to get the logging level you want.

Also, the same verbosity option is currently overloaded for tool use and logging level. I'm thinking eventually (maybe even 2.1.0 if one of us gets around to it) I'd like something like:
- --verbose is reserved for tool use (no change to option parsing or output object)
- a new --debug=<level> option enables stderr and sets the log level (no bump loop needed)

In other words, something "crm_verify -V" would tell crm_verify to give more info specific to its purpose, whereas "crm_verify --debug info" would turn on logging to stderr at info level.

kgaillot replied:
Just FYI, separately, I'm thinking all of the logging API should be internal (the public APIs should become deprecated wrappers). I don't see a convincing use case for external applications to share the Pacemaker log, so I'd just let them take care of their own logging. And certainly Pacemaker doesn't need to be in the generic logging library business.

clumens replied:
Yeah, I like the sound of all of this.  I haven't dug into the logging stuff very much but it seems pretty obtuse.  There are an awful lot of functions involved.  The crm_bump_log_level behavior is weird, I agree.  I never would have thought of that without studying the comments more closely.

I also agree on keeping the API private and keeping us out of writing a logging library.

clumens commented:
Working on this PR makes me think it would be nice to have a library just for command line tool stuff.  It'd contain the error printing I just added, maybe the logging I just added, all the argument processing, whatever else.  It would entirely be private API so we could lose the _internal parts on the headers, not worry about double underbars in function names, etc.  I'm sure this has come up before.

kgaillot commented:
> Working on this PR makes me think it would be nice to have a library just for command line tool stuff. It'd contain the error printing I just added, maybe the logging I just added, all the argument processing, whatever else. It would entirely be private API so we could lose the _internal parts on the headers, not worry about double underbars in function names, etc. I'm sure this has come up before.

Yep that was the original idea behind libpacemaker-internal (which evolved into libpacemaker), but in practice public and private APIs are so intertwined it's not worth it. In the distant future when everything is combined into libpacemaker and tons of public APIs have been deprecated and removed, it might be easier to separate some internal-only stuff, but it's not worth caring about now.

kgaillot commented on include/crm/common/errors_internal.h:
```diff
@@ -0,0 +1,36 @@
+ *
+ * \note \p error and \p out can both be NULL.
+ */
+void pcmk__cli_handle_error(GError *error, pcmk__output_t *out);
```
I would just throw this in output_internal.h rather than add another header, since I don't see us adding much other error-related stuff, and this is output-related even if it doesn't require pcmk__output_t.

Also, I think the name could be misleading since the error is just output, not truly handled. It may only be used by CLI, but I think it's more generic itself. I'm thinking something like pcmk__output_and_free_error() to additionally indicate that error can't be used past that.

kgaillot commented on include/crm/common/errors_internal.h:
```diff
@@ -0,0 +1,36 @@
+ * \param[in,out] error A GError object potentially containing some error.
+ * \param[in]     out   The output functions structure.
+ *
+ * \note \p error and \p out can both be NULL.
```
Rather than this note, I'd document what NULL does by each parameter

clumens commented:
> Yep that was the original idea behind libpacemaker-internal (which evolved into libpacemaker), but in practice public and private APIs are so intertwined it's not worth it. In the distant future when everything is combined into libpacemaker and tons of public APIs have been deprecated and removed, it might be easier to separate some internal-only stuff, but it's not worth caring about now.

I feel like libpacemaker is more the guts of the command line tools, and I was thinking more along the lines of the boring support stuff, as a way of trying to make it faster to write command line tools in the future and keep them looking more consistent.  But yeah, we do have enough problems keeping libraries organized already without adding more.

kgaillot commented:
I don't know if you want to include these in this PR, but issues that came up elsewhere:
* crm_mon --daemonize silently converts text/console output format to none. This is likely because these are the default if no format is specified, and with --daemonize letting the format default is more likely to mean that --external-agent is what is desired. However someone could reasonably want --daemonize with text output, so a better approach would be to convert to none if no --output-to has been specified (regardless of original format).
* print_as() violates the spirit of the output model. It looks like the uses that don't make sense in HTML could simply use ->info() already. Anything that makes sense in HTML could maybe be a new custom message, or since it's already ignored for HTML, we could just use info() and forget about it.
* The crm_mon connection code could be functionized, and we maybe don't need cib and fencer connections to be opened/closed together anymore.

kgaillot commented:
ppc64le build issue looks interesting :)

clumens commented:
> ppc64le build issue looks interesting :)

Heh, those keep sneaking through.  Okay there shouldn't be any more uses of crm_log_cli_init now.

clumens commented:
> I don't know if you want to include these in this PR, but issues that came up elsewhere:
> 
>     * crm_mon --daemonize silently converts text/console output format to none. This is likely because these are the default if no format is specified, and with --daemonize letting the format default is more likely to mean that --external-agent is what is desired. However someone could reasonably want --daemonize with text output, so a better approach would be to convert to none if no --output-to has been specified (regardless of original format).
> 
>     * print_as() violates the spirit of the output model. It looks like the uses that don't make sense in HTML could simply use ->info() already. Anything that makes sense in HTML could maybe be a new custom message, or since it's already ignored for HTML, we could just use info() and forget about it.
> 
>     * The crm_mon connection code could be functionized, and we maybe don't need cib and fencer connections to be opened/closed together anymore.

These can be handled in a separate PR, since they are focused just on crm_mon.  Was Klaus going to look at any of these, or should I just jump right in?

kgaillot commented:
> These can be handled in a separate PR, since they are focused just on crm_mon. Was Klaus going to look at any of these, or should I just jump right in?

Go ahead and take them
 
> Related, what to do about print_dot?

I think that could potentially be useful more broadly, maybe a new out->progress(out) that's essentially out->info(out, ".") without a newline. I don't see anything similar elsewhere but I wouldn't be surprised if it came up at some point. Not long ago we replaced something like that with "... got reply" in crm_resource, I think just because it was easier with the new output model, but this would let us go back if desired. It could also make it easy to change it to a spinning wheel or whatnot in the future as well.

kgaillot commented on include/crm/common/logging_internal.h:
```diff
@@ -0,0 +1,24 @@
+ * \note \p verbosity is not the same as the logging level (LOG_ERR, etc.).
+ */
+void pcmk__cli_init_logging(const char *name, unsigned int verbosity);
+
```
There's an "internal logging utilities" section of include/crm/common/internal.h that could move here (creating logging_internal.h with those could be a separate commit, to make this one cleaner).

Also, this commit message needs the new function name.

clumens replied:
Which utilities are these?  I only see the deprecated block at the bottom.

kgaillot commented on daemons/execd/cts-exec-helper.c:
```diff
@@ -527,7 +527,7 @@ main(int argc, char **argv)
     crm_trigger_t *trig;
 
-    crm_log_cli_init("cts-exec-helper");
+    pcmk__cli_init_logging("cts-exec-helper", 0);
```
This commit looks great, but change the commit message like "API: libcrmcommon: ...". With the new changelog section, that will look like:

    - Public API changes since 2.0.5
      + libcrmcommon: ...
      + ...

kgaillot reviewed (commented):
Looks great, just a couple of nitpicks

kgaillot commented on lib/common/logging.c:
```diff
@@ -978,10 +972,24 @@ crm_log_output_fn(const char *file, const char *function, int line, int level, c
+gboolean crm_log_cli_init(const char *entity);
+
+/*!
+ * \deprecated Use pcmk__cli_init_logging instead
```
Actually we should say something like "Do not use Pacemaker for logging" -- this is the public API documentation, and external code won't have access to pcmk__cli_init_logging(). Eventually we'll deprecate the entire public logging API.

kgaillot replied:
Also we only need the doxygen block in one place (logging.h or logging.c)

clumens replied:
Updated in the header file.
